src/hardware/mixer.cpp

```diff
@@ -633,34 +633,41 @@ void MixerChannel::Enable(const bool should_enable)
 
 void MixerChannel::ConfigureResampler()
 {
-	const auto in_rate = (resample_method == ResampleMethod::ZeroOrderHoldAndResample)
-	                           ? zoh_upsampler.target_freq
-	                           : sample_rate;
-
-	const auto out_rate = mixer.sample_rate.load();
-
-	do_resample = (in_rate != out_rate);
+	const auto channel_rate = sample_rate;
+	const auto mixer_rate   = mixer.sample_rate.load();
 
-	if (!do_resample) {
-		// DEBUG_LOG_MSG("%s: Resampler is off", name.c_str());
-		return;
-	}
+	do_resample = false;
+	do_zoh_upsample = false;
 
 	switch (resample_method) {
 	case ResampleMethod::LinearInterpolation:
-		InitLerpUpsamplerState();
-		// DEBUG_LOG_MSG("%s: Linear interpolation resampler is on",
-		//               name.c_str());
+		do_resample = (channel_rate != mixer_rate);
+		if (do_resample) {
+			InitLerpUpsamplerState();
+			// DEBUG_LOG_MSG("%s: Linear interpolation resampler is on",
+			//               name.c_str());
+		}
 		break;
 
 	case ResampleMethod::ZeroOrderHoldAndResample:
-		InitZohUpsamplerState();
-		// DEBUG_LOG_MSG("%s: Zero-order-hold + Speex resampler is on, upsampler target freq : % d Hz ",
-		//               name.c_str(),
-		//               zoh_upsampler.target_freq);
+		do_zoh_upsample = (channel_rate != zoh_upsampler.target_freq);
```

Comment:
In Disney and the Tandy DAC, the `zoh_upsampler.target_freq` to set to match `channel_rate`, which might appear to actually disable `do_zoh_upsample`, when reading this logic here.

We should clarify (maybe a comment here?) that user-side code needs to: 
1. Set the ZoH rate to the channel rate
2. Change the channel rate
3. Run ConfigureResampler() once the channel rate differs from the ZoH rate.

Reply:
If ZoH is only an upsampler (and can't drop samples), then maybe:

`do_zoh_upsample = (channel_rate < zoh_upsampler.target_freq);`

In the cases where someone is using a mixer rate lower than the upsampling rate, we'll still fallthrough and speex will perform the downsampling.

Reply:
> Suggest a comment stressing that the order of operations are critical to properly setup the ZoH upsampler:
> 
>     1. Set the ZoH rate (typically to the channel rate)
>     2. Change the channel rate
>     3. Run ConfigureResampler() once the channel rate differs from the ZoH rate.

`ConfigureResampler` is called at the end of `SetSampleRate`, `SetZeroOrderHoldUpsamplerTargetFreq` and `SetResampleMethod` to ensure everything is set up correctly when any of these methods get called, regardless of the ordering. Note that this is important at least for the SoundBlaster emulation as games can set the sample rate dynamically, and in some cases do so quite frequently (e.g. in Alone in the Dark there is a `SetSampleRate` call before _every single individual sound effect_ because the game varies the rate slightly to add more variety to the sampled sounds).

Therefore, `ConfigureResampler` has been designed to do the correct thing in all these scenarios. Also note that `ZeroOrderHoldAndResample` means that that we could be using either:

a) no ZoH and no Speex resampling (when `channel_rate == zoh_target_freq == mixer_rate`)
b) only ZoH (when `channel_rate != zoh_target_freq` and `zoh_target_freq == mixer_rate`)
c) only Speex (when `channel_rate == zoh_target_freq`, and `zoh_target_freq != mixer_rate`)

Reply:
> If ZoH is only an upsampler (and can't drop samples), then maybe:
> 
> `do_zoh_upsample = (channel_rate < zoh_upsampler.target_freq);`
> 
> In the cases where someone is using a mixer rate lower than the upsampling rate, we'll still fallthrough and speex will perform the downsampling.

Yes, it can only upsample, and that's a good idea.

Reply:
Add explanatory comment as well.

Reply:
Robust logic and thoroughly documented (in the key areas!) -- Nice, @johnnovak.